include/ppx/metrics.h

```diff
@@ -0,0 +1,198 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_metrics_h
+#define ppx_metrics_h
+
+#include "ppx/config.h"
+
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ppx {
+namespace metrics {
+
+#define METRICS_NO_COPY(type)   \
+    type(type&)       = delete; \
+    type(const type&) = delete; \
+    type& operator=(const type&) = delete;
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricInterpretation
+{
+    NONE,
+    HIGHER_IS_BETTER,
+    LOWER_IS_BETTER,
+};
+
+struct Range
+{
+    double lowerBound = std::numeric_limits<double>::min();
+    double upperBound = std::numeric_limits<double>::max();
+};
+
+struct MetricMetadata
+{
+    std::string          name;
+    std::string          unit;
+    MetricInterpretation interpretation = MetricInterpretation::NONE;
+    Range                expectedRange;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricType
+{
+    GAUGE,
+    COUNTER,
+};
+
+class Metric
+{
+    friend class Run;
+
+public:
+    const MetricMetadata& GetMetadata() const;
+    MetricType            GetType() const;
+
+protected:
+    Metric(const MetricMetadata& metadata, MetricType type);
+    virtual ~Metric();
+
+private:
+    MetricMetadata mMetadata;
+    MetricType     mType;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+struct GaugeStatistics
+{
+    double min;
+    double max;
+    double average;
+    double median;
+    double standardDeviation;
+    double timeRatio;
+    double percentile90;
+    double percentile95;
+    double percentile99;
+};
+
+class MetricGauge final : public Metric
+{
+public:
+    MetricGauge(const MetricMetadata& metadata);
+
+    void   RecordEntry(double seconds, double value);
+    size_t GetEntriesCount() const;
+    void   GetEntry(size_t index, double& seconds, double& value) const;
+
+    const GaugeStatistics GetStatistics(bool realtime) const;
+
+private:
+    struct TimeSeriesEntry
+    {
+        double seconds;
+        double value;
+    };
+
+private:
+    ~MetricGauge();
+    METRICS_NO_COPY(MetricGauge)
+
+    void UpdateRealTimeStatistics(double seconds, double value);
+
+private:
+    std::vector<TimeSeriesEntry> mTimeSeries;
+    GaugeStatistics              mRealTimeStatistics;
+    double                       mAccumlatedValue;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class MetricCounter final : public Metric
```

Comment:
Comments? Assuming you follow the same principle as https://prometheus.io/docs/concepts/metric_types/ ?

Reply:
I added comments in the metrics.h to explain what the metrics do, among other comments.